rfd/0030-upgrade-system.md

```diff
@@ -0,0 +1,444 @@
+index/manifest system will continue to provide checksums with correct versioning information.  Some
+repository systems also resist tampering with version numbers, though not all do so.  That leaves us
+with one final scenario: the index *and* at least one repository are compromised simultaneously (but
+the compromise is not so severe as to put us in the previously discussed 'doomsday scenario').  This is
+a bit niche, but lets exercise a little professional paranoia.  We should have a mechanism of validating
+package version (prior to unpacking/installation) as part of package signature verification (i.e. on the
+"distribution" side).  One appealing option is [minisign](https://jedisct1.github.io/minisign/) which
+is a simple and modern signing utility that supports "trusted comments" (i.e. signed metadata) as part
+of the signature file.
+
+
+#### Compromised Build System
+
+Out of scope for now, but I've heard all the cool kids are doing deterministic builds these days.
+
+
+## Proposal
+
+### Overview
+
+High-level overview of system components, including discussion of MVP vs planned features.
+
+#### Index Server
+
+The index server, as discussed above, is responsible for serving metadata and checksums for the
+current list of valid installation targets.  Part of the security model of the upgrade system is
+that the index server remains isolated from the rest of the ugrade system stack, and provides only
+abstract information about available releases (i.e. it will not provide things like download urls
+or install scripts).
+
+Key design considerations related to the index server are as follows:
+- Index server should be trivially horizontally scalable.
+- Index servers should always require TLS 1.3 unless started in development mode.
+- Rollbacks of index server state or API version must not be detremental to its clients
+(i.e. clients must be resilient to inconsistent values being returned).
+- Index server must remain compatible with many past teleport versions (exact number TBD),
+but we don't want to orphan older versions, so lets err on the side of caution here.
+
+Outstanding design questions:
+- Grpc or basic https API?  My preference leans toward grpc, but its hard to discount the possible
+benefits of a basic https API (e.g. pulling down release info with `curl`).
+- How 'smart' should the index server be exactly?  On one hand, if the index server is able to intelligently
+select what releases/targets are relevant to a given client, it may be able to reduce the impact of bugs
+in the upgrade system (easier to update the index server than to update all clients).  On the other hand,
+the extensibility of the upgrade system (not to mention the simplicity of the index server) would likely
+be benifited by keeping as much logic as possible in the client clusters.
+
+Hypothetical teleport release descriptors returned by index server:
+
+```yaml
```

Comment:
this reminds me of https://certificate.transparency.dev/ 

keeping this log tamper-proof and easilly verifiable is important.

another consideration is that we don't implement lots of code here, for example seeing if we can leverage https://github.com/google/trillian with RDS backed MySQL would reduce the amount of work hopefully

Reply:
One thought about this index and it's context. Relying on TLS only won't let users to validate the changes to the index contents, while signatures on it can. Probably that's why indexes are signed, what adds additional layer of verification.

Customers can also copy indexes and host their mirrors to avoid reaching out to internet, similar to package contents.

Reply:
IMO the linked solutions are too heavyweight for this application.  Merkle trees are cool, but they aren't really useful unless you're dealing with proving membership in a very large dataset to a client that cannot practically download the dataset itself (or in the case of things like torrents, may download the dataset slowly in small chunks from various sources).  IMO a list of release manifests, even *many years* of release manifests, is far too small a dataset to bother with merkle proofs.

I'm also not convinced on the idea that we need to care about making release manifests "tamper proof" in the sense meant by the linked projects (I'd love to have my mind changed on this tho, tamper proof datastructures are cool!).  IIUC the tamper proofing offered by the above projects relies upon some combination of "pinning" (e.g. client is pre-configured with some merkle root in the not-too-distant past to cross-check against), and an ecosystem of public validators which can keep each other honest about more recent entries.  The "pinning" of old state isn't super helpful for an upgrade system because we likely care about, at most, one version previous.  The ecosystem of public validators half is a bit optimistic.  Unless multiple broadly trusted organizations make a long-term commitment to running public validator endpoints, it isn't helpful.  Clients that don't cross-check with various independent sources (or are willing to skip cross-checking when those sources are unavailable) aren't meaningfully more secure than clients that just trust the one source (again, change my mind please!).

If we do decide to make the log tamper-proof, a simpler way of doing so would be to include a nonce, and the hash of the previous entry (release manifest) in the body of each log entry.  No merkle proofs, but no complicated dependencies either.  Anyone could write a validator in a few minutes.

Reply:
Yea, I would be cautious about trying to use trillian as well, but for different reasons. I've been keeping an eye on some of the attempts to use this for binary transparency which is where this really comes from, and I'm starting to get a feeling it's sort of a flawed idea when ported to binary transparency. Also you would want another company to run the transparency server, doesn't do much if it's the same company providing both.

Reply:
Ok, fair points re: complexity and infrastructure. What if we simply publish index on github for example, and sign every single addition + all the versions using some simple tool and publish the public key on our own repo for anyone to validate?